source/basic.tex

```diff
@@ -5383,6 +5383,8 @@
 after the end of the storage occupied by the object,
 respectively.
 \begin{note}
+A pointer to an actual array element object is categorized as a pointer to an object,
+even though it also points past the end of the previous array element, if any.
```

Comment:
Do not get what does
>  it also points past the end of the previous array element

mean.

Reply:
> > I think this is as clear (or unclear) as "a pointer past the end of an object"
> 
> You mean a pointer to an array element is also a pointer past the end of the previous array element?

I think we can surely say a pointer that is "a pointer past the end of an object" points past the end of some given object. (Isn't this a tautology?) And then, given `T a[2]`, there're two choices:
1. there's a pointer value pointing past the end of `a[0]`, and if so, `a + 1` is exactly that value; or
2. there's no pointer value pointing past the end of `a[0]`.

Personally, I think choice 1 makes a little more sense. But if you think choice 2 makes more sense, we may abandon this PR and clarify in that choice.

Reply:
> According to [expr.add], `a + 1` points to 1st array element. It can't be pointer past 0th at the same time, because [Every value of pointer type is **one of** the following](https://timsong-cpp.github.io/cppwp/n4868/basic.compound#3.sentence-8)

Well, I think we agree on that `a + 1` is categorized as a pointer to an object, and I don't think `a + 1` is the "pointer past 0th" (Isn't this the content of this PR?).

What I wanted to clarify is that `a + 1` points past the end of `a[0]` without being the pointer past the end of `a[0]`.

Reply:
> `a + 1` points past the end of `a[0]`

And this is my question: what does this mean

Reply:
> And this is my question: what does this mean

IMO when `p` points to an object `o`, we can generally say `p + 1` points past the end of `o` (perhaps some base class subobjects should be excluded).